### PR TITLE
Allow spaces when autocompleting mentions

### PR DESCRIPTION
**Pull request by danxuliu**
By default Tribute does not allow spaces when autocompleting mentions, so as soon as a space was pressed the autocompletion list was closed, even if some candidate mentions had a space (either in the id or in the display name of the user).

As I found no explicit mention about why spaces should not be allowed when autocompleting mentions in `NcRichContenteditable` this pull request enables them (but if there is a good reason not to enable them please shout :-) ).

Note that now the autocompletion does not end until _Enter_ is pressed (or an item is selected in the list); even if space is pressed and nothing else matches the autocomplete will still be active, even if the list is hidden.

This pull request also fixes an existing bug: if _Enter_ was pressed when the current mention did not match anything an error was shown in the console; now the current unmatched mention text is passed.

As the Comments app in Nextcloud server uses `NcRichContenteditable` these issues are also present in previous Nextcloud server releases. But I do not know how that is handled :shrug: Is this pull request backported to previous nextcloud-vue versions, a bugfix release is published for those older versions, and then the dependencies are bumped to the respective bugfix releases in _stable23_, _stable24_ and _stable25_ in server?

## How to test (scenario 1)

- Open the `NcRichContenteditable` documentation ([master](https://nextcloud-vue-components.netlify.app/#/Components/NcRichContenteditable), [pull request](https://deploy-preview-3468--nextcloud-vue-components.netlify.app/#/Components/NcRichContenteditable))
- In the example, type `@test 03`

### Result with this pull request

The autocompletion list is kept open and only `Test 03` is shown

### Result without this pull request

The autocompletion list is closed

## How to test (scenario 2)

- Open the `NcRichContenteditable` documentation ([master](https://nextcloud-vue-components.netlify.app/#/Components/NcRichContenteditable), [pull request](https://deploy-preview-3468--nextcloud-vue-components.netlify.app/#/Components/NcRichContenteditable))
- In the example, type `@unmatched`, press _Enter_ and then press _Enter_ again

### Result with this pull request

The message is sent as `@unmatched`

### Result without this pull request

An error is shown twice in the console and the message is not sent

Note that, with this pull request, if `@Test 03 and some more text` is written and _Enter_ is pressed twice then the message sent will be `@Test 03 and some more text` rather than `@"Test 03" and some more text`, given that as mentioned above the autocompletion does not end until _Enter_ is pressed, so `@Test 03 and some more text` is seen as the whole unmatched mention and therefore passed. I have not found a way to solve that, although it would be probably uncommon to type the full user id without user the autocompletion list. Or at least I hope so :-)

### CODE REVIEW

**Comment by skjnldsv** [comment]
> even if some candidate mentions had a space (either in the id or in the display name of the user).

Spaces are allowed in user IDs?

**Comment by danxuliu** [comment]
> > even if some candidate mentions had a space (either in the id or in the display name of the user).
> 
> Spaces are allowed in user IDs?

[Yes, they are](https://github.com/nextcloud/server/blob/8e8acf2d9087523ac8f8bf3aeac8daa0c17add91/lib/private/User/Manager.php#L432) :shrug:

File: src/mixins/richEditor/index.js [hunk 1]
```diff
@@ -104,6 +104,10 @@ export default {
 		 * @return {string}
 		 */
 		genSelectTemplate(value) {
+			if (typeof value === 'undefined') {
+				return `${this.autocompleteTribute.current.collection.trigger}${this.autocompleteTribute.current.mentionText}`
```
Comment (artonge) [inline src/mixins/richEditor/index.js:108]: Can you add a comment saying something like:
> // What's returned by this function will replace the trigger '@' and the search text, so when there is no match, then we simply return '@' and the text

**Review by artonge — COMMENTED** [review]
I am fine with the code, but I think we can take the opportunity to improve the flow a little bit.

Currently, when there is no match, the pop-up is not displayed, so the user does not know that he is still triggering tribute. Which means that when he presses enter, it will trigger tribute without knowing it.

I think it is best to keep the popup displayed, so the user is not lost.

Achievable with something like:

```js
noMatchTemplate: `<div>${translate("No comments")}</div>`
```

@jancborchardt or @nimishavijay

File: src/mixins/richEditor/index.js (hunk 1 above)
Comment (danxuliu) [reply src/mixins/richEditor/index.js:111]: Done :-)

**Review by artonge — APPROVED** [review]

**Comment by szaimen** [comment]
> > > even if some candidate mentions had a space (either in the id or in the display name of the user).
> > 
> > 
> > Spaces are allowed in user IDs?
> 
> [Yes, they are](https://github.com/nextcloud/server/blob/8e8acf2d9087523ac8f8bf3aeac8daa0c17add91/lib/private/User/Manager.php#L432) 🤷

There seems to be an issue within the regex. Shouldnt we fix this in server? Cc @nickvergessen

**Comment by nickvergessen** [comment]
What is the issue? Spaces are allowed and always have been. So need to continue to work

**Comment by szaimen** [comment]
> What is the issue? Spaces are allowed and always have been. So need to continue to work

Ah I see. Only asking because the comment says otherwise IIRC

**Comment by nickvergessen** [comment]
https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/user_configuration.html#creating-a-new-user mentions spaces, so yeah fix the code comment and the error message, not the regex.

**Comment by szaimen** [comment]
> https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/user_configuration.html#creating-a-new-user mentions spaces, so yeah fix the code comment and the error message, not the regex.

done in https://github.com/nextcloud/server/pull/35363. I somehow doubt that this is really correct but fine by me if you say so.

**Comment by nickvergessen** [comment]
> I somehow doubt that this is really correct but fine by me if you say so.

There are instances with spaces in user ids, we fixed X issues with the URL encoding not working fully in dav over the years.

**Review by marcoambrosini — APPROVED** [review]

**Pull request merged**